Feat 498 pyd v2

jtyoung84 opened the pull request:
Closes #498

## Still in progress

### Updating the validators
- v2 removes `validator` and `root_validator` and changes them to `field_validator` and `model_validator` with different interfaces

### Resolving serialization warnings
- There seems to be an issues doing round trip serialization/de-serialization of nested models. I'm looking into this issue further.

## Upgrades pydantic to version 2

### Changes in the example json files:
- Decimal class now gets rendered as strings instead of floats. (This change is unavoidable)
- I changed stuff like `Optional[List[str]] = None` to `List[str] = []` and changed the corresponding null values to empty lists. I can change them back, but this makes things slightly easier to maintain since we don't need to do checks against None. Again, if this isn't the desired behavior, I can change things back
- Everything else remains the same, so from a user standpoint, it shouldn't be too different

### Changes to the schemas
- It seems v2 renders fields like `field_name: Optional[str]` as `anyOf` in the schema. I think in v1 it just marked the field as `"requried": false`. I'm not sure if there's a way to prevent that change.
- This is a work in progress, but it's been difficult to use non-string and non-int enums. I changed all of them to "tagged" classes. For example, modalities and institutions are different. I'm working on making it so the users can still use the same interface (modality=Modality.ECEPHYS), but I need to do some refactoring. For the standard str and int enums, I kept them the same. I set the base model config to use_enum_values=True so the schema just renders the value and not the class name.
- pydantic raises warnings about adding extra information in the fields outside of the description. I removed all instances of "units=" and "additionalProperties=".
- I'll need to explore more how different the changes are between the old v1 and v2 schemas.
- I changed some of the class names that were the same across modules (session.Laser and device.Laser for example). I renamed the stuff in session to something like `LaserConfigs` but I can rename them back to their original name, but it was a bit confusing knowing which was which.

### Changes to the package organization
- I removed the deprecated packages
- I put all the core models into a single package to make it easier to know which classes are AindCoreModels
- I put models that get imported into different packages into a models package. This makes it easier to avoid circular imports and having to use future annotations.

saskiad commented on examples/aibs_smartspim_instrument.json:
```diff
@@ -25,12 +25,13 @@
          "notes": null,
-         "length": 35,
-         "width": 29,
+         "length": "35",
```
are all number fields now strings?

jtyoung84 replied:
The Decimal types are now serialized to strings when converted to json in pydantic v2. There's a lengthy discussion about it in this thread: https://github.com/pydantic/pydantic/issues/1511

saskiad commented on src/aind_data_schema/core/acquisition.py:
```diff
@@ -51,26 +49,27 @@ class Axis(AindModel):
-            axes.append(axis)
-
-        return axes
+    # #  TODO: It might make more sense to put this in the utils package
```
did you move it to utils? or just remove it?

jtyoung84 replied:
I commented it out currently and haven't decided where it should go. There can be some issues self-referencing the class from a staticmethod. I think I'll probably end up moving the function into the utils package.

jtyoung84 replied:
I ended up adding it as a field_validator on the Acquisition class. A user can either input a list of Axis objects or a string of direction codes:
https://github.com/AllenNeuralDynamics/aind-data-schema/blob/c3a9c8d95525cc14ab8071750dbd986bb4fd1df5/src/aind_data_schema/core/acquisition.py#L118

saskiad commented on src/aind_data_schema/core/acquisition.py:
```diff
@@ -111,13 +113,13 @@ class Acquisition(AindCoreModel):
     instrument_id: str = Field(..., title="Instrument ID")
-    calibrations: Optional[List[Calibration]] = Field(
-        None,
+    calibrations: List[Calibration] = Field(
```
it's unclear to me why making Optional fields required with default of an empty list is preferred over having them be optional. Is there a reason?

jtyoung84 replied:
I'm on the fence about this change. I can revert it back if it's preferred. It's slightly easier to maintain using an empty list as a default, but I can understand why there might be a meaningful difference between "this field has no input" vs. "this field is an empty list"

saskiad replied:
part of my concern is how users understand what things are required and what are optional. I think this syntax is going to be too subtle.

jtyoung84 replied:
I can change it to something like calibrations: List[Calibrations] = Field(default = [], etc.) so that users know that there is a default value. Pydantic interprets `Optional` as something like a nullable field

saskiad commented on src/aind_data_schema/core/instrument.py:
```diff
@@ -0,0 +1,96 @@
+
+from aind_data_schema.base import AindCoreModel, AindModel
+from aind_data_schema.models.devices import (
+    LIGHT_SOURCES,
```
why is this in caps?

jtyoung84 replied:
Light sources isn't a class, it's a constant referencing a list of classes a user can select from.

saskiad commented on src/aind_data_schema/core/instrument.py:
```diff
@@ -0,0 +1,96 @@
+    com_ports: List[Com] = Field([], title="COM ports")
+    notes: Optional[str] = None
+
+    # @root_validator
```
we're going to replace this, right?

jtyoung84 replied:
Yeah, I'm going to refactor the validators last. They are going to require the most work since pydantic v2 has a completely different interface.

jtyoung84 commented on src/aind_data_schema/core/mri_session.py:
```diff
@@ -2,56 +2,17 @@
+from aind_data_schema.core.acquisition import Axis
+from aind_data_schema.core.procedures import Anaesthetic
 from aind_data_schema.imaging.tile import Scale3dTransform
-from aind_data_schema.procedures import Anaesthetic
```
I moved the core models into a core package, so this import is now on line 11 as `from aind_data_schema.core.procedures import Anaesthetic`

dyf reviewed (approved):
There are a few things we need to clean up still (notably, the syntax for specifying PIDNames), but I think this is fine to merge.

yosefmaru commented on examples/ephys_session.json:
```diff
@@ -83,38 +85,36 @@
    "reward_consumed_unit": "microliter",
+   "reward_delivery": null,
    "notes": null
 }
```
EOF blank line is missing.

yosefmaru commented on examples/exaspim_acquisition.json:
```diff
@@ -162,12 +162,12 @@
-   "processing_steps": null,
+   "processing_steps": [],
    "notes": null
 }
```
EOF blank line

yosefmaru commented on examples/exaspim_instrument.json:
```diff
@@ -414,5 +413,6 @@
-   "notes": null
+   "notes": null,
+   "enclosure": null
 }
```
EOF blank line

jtyoung84 commented:
@mekhlakapoor @yosefmaru We can ignore github's new line warnings for the auto-generated json files

saskiad commented on src/aind_data_schema/core/mri_session.py:
```diff
@@ -2,56 +2,17 @@
-    scanner_location: ScannerLocation = Field(..., title="Scanner location")
-    magnetic_strength: MagneticStrength = Field(..., title="Magnetic strength (T)", units="T")
-    magnetic_strength_unit: str = Field("T", title="Magnetic strength unit")
+from aind_data_schema.models.devices import MriScanSequence, Scanner, ScanType
```
I'm not a fan of moving these to devices because they aren't devices.

jtyoung84 replied:
I can move the MriScanSequence and ScanType classes to the mri_session module. The Scanner inherits from the Device class though.

saskiad commented on src/aind_data_schema/models/devices.py:
```diff
@@ -305,104 +421,75 @@ class DAQChannel(AindModel):
     # required fields
-    device_type: Literal["Harp device"] = Field("Harp device", const=True, readOnly=True)
+    device_type: Literal["Harp device"] = "Harp device"
     harp_device_type: HarpDeviceType = Field(..., title="Type of Harp device")
```
I'm sure it's somewhere, but where is the enum of HarpDeviceType living now? I hvaen't been able to find it (probably how the diff gets shown) but the first place I'd look is with the Harp Device class, so it makes me uncomfortable having it not easily findeable.

jtyoung84 replied:
I put all of the enum definitions at the top of the devices.py file to avoid potential forward reference conflicts. We can re-organize them to different places though if it's desirable.

saskiad commented on src/aind_data_schema/core/procedures.py:
```diff
@@ -334,11 +332,11 @@ class InjectionMaterial(AindModel):
     genome_copy: Optional[Decimal] = Field(None, title="Genome copy")
     titer: Optional[Decimal] = Field(None, title="Titer (gc/mL)", description="Titer for viral materials")
-    titer_unit: Optional[str] = Field("gc/mL", title="Titer unit")
+    titer_unit: Literal["gc/mL"] = "gc/mL"
```
titer might be the exception to this rule though, so we can leave it here, but in general

jtyoung84 replied:
I'll revert it back to an Optional[str]

saskiad commented on src/aind_data_schema/core/procedures.py:
```diff
@@ -350,29 +348,29 @@ class InjectionMaterial(AindModel):
-    injection_materials: List[InjectionMaterial] = Field(
-        None, title="Injection material", unique_items=True, min_items=1
-    )
+    procedure_type: Literal["Injection"] = Field("Injection", title="Procedure type")
```
Injection is never used without being used as part of a different procedure type (Nanoject or Iontophoresis or Retro-Orbital). We want those procedure types to be what is recorded and not "Injection".

jtyoung84 replied:
I'll remove the field here

saskiad commented on src/aind_data_schema/core/procedures.py:
```diff
@@ -350,29 +348,29 @@ class InjectionMaterial(AindModel):
 class BrainInjection(Injection):
     """Description of a brain injection procedure"""
 
+    procedure_type: Literal["BrainInjection"] = Field("BrainInjection", title="Procedure type")
```
Ditto above - the procedure type should not be Brain Injection

jtyoung84 replied:
I'll remove the field here

saskiad commented on src/aind_data_schema/core/procedures.py:
```diff
@@ -98,12 +76,94 @@ class Fluorophore(Enum):
+class SpecimenProcedure(AindModel):
+    """Description of surgical or other procedure performed on a specimen"""
+
+    specimen_procedure_type: Literal["SpecimenProcedure"] = Field(
```
wait, we agreed to get rid of specimen_procedure_type, correct? We can do that later, but just so we're on the same page.

jtyoung84 replied:
I think since this is how it is in the main branch, it might be better to update it in a future PR.

saskiad commented on src/aind_data_schema/models/species.py:
```diff
@@ -0,0 +1,70 @@
+    HOMO_SAPIENS = HomoSapiens()
+    MACACA_MULATTA = MacacaMulatta()
+    MUS_MUSCULUS = MusMusculus()
+    ALL = tuple(_Species.__subclasses__())
```
Can someone explain how the ALL thing works. We don't want people to choose ALL in these enums. I assume it is so we can use the list of enums somehow, but isn't that what the enum does in the first place?

jtyoung84 replied:
I imagine we'll need to revisit in a future PR how we want to manage the situation where users can select from a list of non-primitive types. Pydantic doesn't make it too easy. I can add an underscore to the `ALL` to mark it as a protected attribute that won't be exposed to the user. I added it to the classes where a user needs to select OneOf different values. For classes where users don't need to make a selection like that, I didn't include it.

jtyoung84 replied:
Also, most of the threads I've read recommended using tagged Unions instead of enums for non-primitive types. In pydantic v1, I added some hacks to the Enum class to get it to work, but I changed things to use tagged Unions for users to select a value from, but that also requires some additional work on our end.

saskiad commented on src/aind_data_schema/core/session.py:
```diff
@@ -159,45 +158,48 @@ class ManipulatorModule(DomeModule):
     """Description of laser settings in a session"""
 
-    device_type: Literal["Laser"] = Field("Laser", const=True, readOnly=True)
+    config_type: Literal["LaserConfig"] = "LaserConfig"
```
I think we want to keep this as device_type. It's to indicate what type of lightsource is being used. Eg. the name needs to match a name of a laser in the rig schema.

saskiad replied:
ditto for the LightEmittingDiode

jtyoung84 replied:
I'll revert the field back to the original

saskiad commented on src/aind_data_schema/core/session.py:
```diff
@@ -45,10 +44,10 @@ class Detector(AindModel):
     """Description of LED settings"""
 
-    device_type: Literal["LightEmittingDiode"] = Field("LightEmittingDiode", const=True, readOnly=True)
+    config_type: Literal["LightEmittingDiodeConfig"] = "LightEmittingDiodeConfig"
```
see comment in LaserConfig - we want to stick with device_type here

jtyoung84 replied:
I'll revert the field back to the original

saskiad commented:
I've left some comments - take a look. If you want to ignore them for now and turn them into issues to fix, that's fine. They're mostly small. A few might be worth fixing now. But if you do merge, let's turn them into issues so we fix them soon.

jtyoung84 commented on src/aind_data_schema/models/manufacturers.py:
```diff
@@ -0,0 +1,620 @@
+    registry_identifier: Literal[None] = Field(None)
+
+
+class CarlZeiss(_Manufacturer):
```
python classes traditionally start with upper case letters

jtyoung84 replied:
Although I can change this value since this is the value that gets exposed to the user:
```
name: Literal["Carl Zeiss"] = "Carl Zeiss"
```
Should that be something else?

jtyoung84 commented:
> I've left some comments - take a look. If you want to ignore them for now and turn them into issues to fix, that's fine. They're mostly small. A few might be worth fixing now. But if you do merge, let's turn them into issues so we fix them soon.

I pushed a couple of commits that addresses your comments
